src/ppx/scene/scene_scene.cpp

```diff
@@ -122,5 +133,50 @@ ppx::Result Scene::AddNode(scene::NodeRef&& node)
     return ppx::SUCCESS;
 }
 
+scene::ResourceArrayIndexMap<scene::Sampler> Scene::GetSamplersArrayIndexMap() const
```

Comment:
I'm not sure I follow, why is this indexing is required?

My shader resources are kept in the resource manager, so unordered map: meaning I don't get to keep the order I added them with.

But here you rebuild the index depending on the map iteration. So this index is in fact not given by the user (depending on the shader binding indices), but computed from the unordered_map.
Hence I'm a bit lost: what is this index map providing?
Also, there is one function to return the vector from the map, so index depends on the map ordering. But here you use this vector, to create another map to explicitly store those indices as value in the map? Is this to do some kind of lookup?

How do you plan to use those?

Reply:
I apologize that this is a bit cryptic, but it should make more sense once the samples are in.  The idea here is to keep the bindings or array offsets agnostic since the scene can have arbitrary number of objects.

Each draw call has a push constants block that looks something like this:
```
struct Draw {
  uint instanceIndex; // This indexes into a StructuredBuffer for modeMatrix
  uint materialIndex; // This indexes into a StructuredBuffer for material params, texture index, sampler in dex
};
```

In the draw loop:
```
for (pNode in DrawableNodes) {
   instanceIndex = NodeIndexMap[pNode];
  materialIndex = MaterialIndexMap[pNode->material];
}
```
`NodeIndexMap` is the `.second` in the returned pair. It maps the pointer to a specifc index in the vector ( `.first` in the returned pair).

When populating the resource for the GPU the vector is walked because it has the index and the pointer to the resource.

Reply:
Right, so you need to have an index<>resource pair kept. And because you need a contiguous `[0;n[` indices ranges at each drawcall, you use this vector: the set for quick insert/removal, the vector to get this contiguous range of indices.
And on top, the map to quickly query which index a value has.

Because the resource manager doesn't handle deletion, this could be simplified with the map + counter. When adding a value, add it to the map, and store the counter as value. Then increment counter.
If you add deletion in the future (not used in BigWheels in any project yet), then it would become more complex. But for now, this seems to be the simples option vs a set, then a vector, then a map no?

Reply:
Was thinking more about this. The array isn't actually needed at all after the map to index is built. The continuous range isn't a strict requirement as long as the highest index is below the max number of descriptor bindings. 

My completely untested worry about the counter is that might overflow the descriptor bindings. 

Regarding deletion, it's not a problem if you just repopulate th descriptor bindings. There's a performance cost, I can only speculate what that is right now would like to measure for sure and find out.